mono/mini/exceptions-amd64.c

```diff
@@ -71,6 +72,125 @@ static LONG CALLBACK seh_unhandled_exception_filter(EXCEPTION_POINTERS* ep)
+
+	/* move current stack into rdi for later restore */
+	amd64_mov_reg_reg (code, AMD64_RDI, AMD64_RSP, 8);
+
+	/* use the new freed stack from sigcontext */
+	amd64_mov_reg_membase (code, AMD64_RSP, AMD64_RCX, (MONO_STRUCT_OFFSET (CONTEXT, Rsp)), 8);
+
+	/* reserve 32 bytes of stack space for Win64 calling convention */
+	amd64_alu_reg_imm (code, X86_SUB, AMD64_RSP, 32);
+
+	/* get the current domain */
+	amd64_mov_reg_imm (code, AMD64_R11, mono_domain_get);
+	amd64_call_reg (code, AMD64_R11);
+
+	/* pop 32 bytes of stack space for Win64 calling convention */
+	amd64_alu_reg_imm (code, X86_ADD, AMD64_RSP, 32);
+
+	/* get stack overflow exception from domain object */
+	amd64_mov_reg_membase (code, AMD64_RAX, AMD64_RAX, G_STRUCT_OFFSET (MonoDomain, stack_overflow_ex), 8);
+
+	/* reserve 32 bytes of stack space for Win64 calling convention */
+	amd64_alu_reg_imm (code, X86_SUB, AMD64_RSP, 32);
+
+	/* call mono_arch_handle_exception (sctx, stack_overflow_exception_obj, FALSE) */
+	amd64_mov_reg_imm (code, AMD64_R8, 0);
+	amd64_mov_reg_reg (code, AMD64_RDX, AMD64_RAX, 8);
+	amd64_mov_reg_reg (code, AMD64_RCX, AMD64_R12, 8);
+	amd64_mov_reg_imm (code, AMD64_R11, mono_arch_handle_exception);
+	amd64_call_reg (code, AMD64_R11);
+
+	/* pop 32 bytes of stack space for Win64 calling convention */
+	amd64_alu_reg_imm (code, X86_ADD, AMD64_RSP, 32);
+
+	/* restore the SEH handler stack */
+	amd64_mov_reg_reg (code, AMD64_RSP, AMD64_RDI, 8);
+
+	/* return */
+	amd64_ret (code);
+
+	g_assert ((code - start) < 128);
+
+	return start;
+}
+
+/* Special hack to workaround the fact that the
+* when the SEH handler is called the stack is
+* to small to recover.
+*
+* Stack walking part of this method is from mono_handle_exception
+*
+* The idea is simple;
+*  - walk the stack to free some space (64k)
+*  - set esp to new stack location
+*  - call mono_arch_handle_exception with stack overflow exception
+*  - set esp to SEH handlers stack
+*  - done
+*/
+static void
+win32_handle_stack_overflow (EXCEPTION_POINTERS* ep, struct sigcontext *sctx)
+{
+	MonoDomain *domain = mono_domain_get ();
+	MonoJitTlsData *jit_tls = mono_tls_get_jit_tls ();
+	MonoLMF *lmf = jit_tls->lmf;
+	MonoContext initial_ctx;
+	MonoContext ctx;
+	guint32 free_stack = 0;
+	StackFrameInfo frame;
+
+	mono_sigctx_to_monoctx (sctx, &ctx);
+
+	initial_ctx = ctx;
+	free_stack = (guint8*)(MONO_CONTEXT_GET_BP (&ctx)) - (guint8*)(MONO_CONTEXT_GET_BP (&initial_ctx));
+
+	/* try to free 64kb from our stack */
+	do {
+		MonoContext new_ctx;
+		MonoJitInfo* ji;
+
+		ji = mono_jit_info_table_find (domain, MONO_CONTEXT_GET_IP (&ctx));
+		mono_arch_unwind_frame (domain, jit_tls, ji, &ctx, &new_ctx, &lmf, NULL, &frame);
+		if (!frame.ji) {
+			g_warning ("Exception inside function without unwind info");
+			g_assert_not_reached ();
+		}
+
+		if (frame.ji != (gpointer)-1) {
+			free_stack = (guint8*)(MONO_CONTEXT_GET_BP (&ctx)) - (guint8*)(MONO_CONTEXT_GET_BP (&initial_ctx));
```

Comment:
What happens if current frame doesn't use frame pointer or use the frame register for other purposes?